tests/test_caplugin.py

```diff
@@ -1,355 +1,167 @@
-        if not results:
-            # First response; now disconnect.
-            results.append(result)
-            ioc.communicate("exit()")
-        else:
-            # Second response; done.
-            results.append(result)
-            break
+    thens = [
+        datetime.fromisoformat(r["time"]["datetime"])
+        for r in result.data["putChannels"]
+    ]
+    now = datetime.now()
+    for then in thens:
+        diff = now - then
+        # Shouldn't take more than this time to get the result of a put out
+        assert diff.total_seconds() < 0.2
 
-    assert len(results) == 2
-    assert results[0] == dict(
-        data=dict(
-            subscribeChannel=dict(value=dict(float=42), status=dict(quality="VALID"))
-        )
-    )
-    assert results[1] == dict(
-        data=dict(subscribeChannel=dict(value=None, status=dict(quality="INVALID")))
+
+@pytest.mark.asyncio
+async def test_put_list(ioc: Popen, schema, list_put):
+    query = list_put
+    result = await schema.execute(query)
+    assert result.data == dict(
+        putChannels=[
+            dict(
+                value=dict(
+                    stringArray=["0.0", "1.7", "2.0"], base64Array=BASE64_0_1688_2
+                )
+            )
+        ]
     )
 
 
 @pytest.mark.asyncio
-async def test_subscribe_ticking(engine: Engine, ioc: Popen):
-    query = (
-        """
-subscription {
-    subscribeChannel(id: "ca://%sticking") {
-        value {
-            string(units: true)
-        }
-        display {
-            precision
-            units
-        }
-    }
-}
-"""
-        % PV_PREFIX
+async def test_put_base64(ioc: Popen, schema, base64_put):
+    # first dump gives {"key": "value"}, a json string
+    # second dump gives \{\"key\": \"value\"\}, an escaped json string
+    query = base64_put
+    result = await schema.execute(query)
+    assert result.data == dict(
+        putChannels=[dict(value=dict(stringArray=["0.0", "1.7", "2.0"]))]
     )
+
+
+@pytest.mark.asyncio
+async def test_subscribe_ticking(ioc: Popen, schema, ticking_subscribe):
+    query = ticking_subscribe
     results = []
-    wait_for_ioc(ioc)
+    await caput(PV_PREFIX + "ticking", 0.0)
     start = time.time()
-    async for result in engine.subscribe(query, context=make_context()):
-        results.append(result)
+    resp = await schema.subscribe(query)
+    async for result in resp:
         if time.time() - start > 0.9:
             break
+        results.append(result.data)
+
     for i in range(3):
         display = None
         if i == 0:
             display = dict(precision=5, units="mm")
         assert results[i] == dict(
-            data=dict(
-                subscribeChannel=dict(value=dict(string="%.5f mm" % i), display=display)
-            )
+            subscribeChannel=dict(value=dict(string="%.5f mm" % i), display=display)
         )
     assert len(results) == 3
 
 
+# !! Must be the last test as it calls a disconnect
```

Comment:
IOC lifecycle management in a test environment is tricky because the processes seem to not always be cleaned up well.

@MJGaughran having worked recently on the topup integration tests, do you have any suggestions for the pattern to use for that?

Reply:
I agree that it is not very nice having to maintain an order to the tests and this would likely lead to maintainability issues so thank you for catching this.
I have extracted some of the methods used to create the session IOC so that I can call them manually to run a separate IOC for just the one test (i.e. the one that tests a disconnect). I purposely changed the test ordering to prove that this works. Please see commit: [69305fb](https://github.com/dls-controls/coniql/pull/44/commits/69305fbf12c553e6c5f97b95f9caafbfb4d33c0b).